Allocation module - Centos 7 Virtualbox and AWS mirror list failure

## Description

`Centos 7` `Virtualbox` and `AWS` mirror list failure

Branches: `4.9.0` and `4495-dtt1-release`

When I tried to install packages by using `yum`, I got the following error
According to a brief search carried out, it could be due to a failure in the centos mirrorlist, but it could also be due to some error that prevents the correct functioning of the DNS.


```
[root@localhost vagrant]# yum -y update
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
Could not retrieve mirrorlist http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=os&infra=vag error was
14: curl#6 - "Could not resolve host: mirrorlist.centos.org; Unknown error"


 One of the configured repositories failed (Unknown),
 and yum doesn't have enough cached data to continue. At this point the only
 safe thing yum can do is fail. There are a few ways to work "fix" this:

     1. Contact the upstream for the repository and get them to fix the problem.

     2. Reconfigure the baseurl/etc. for the repository, to point to a working
        upstream. This is most often useful if you are using a newer
        distribution release than is supported by the repository (and the
        packages for the previous distribution release still work).

     3. Run the command with the repository temporarily disabled
            yum --disablerepo=<repoid> ...

     4. Disable the repository permanently, so yum won't use it by default. Yum
        will then just ignore the repository until you permanently enable it
        again or use --enablerepo for temporary usage:

            yum-config-manager --disable <repoid>
        or
            subscription-manager repos --disable=<repoid>

     5. Configure the failing repository to be skipped, if it is unavailable.
        Note that yum will try to contact the repo. when it runs most commands,
        so will have to try and fail each time (and thus. yum will be be much
        slower). If it is a very temporary problem though, this is often a nice
        compromise:

            yum-config-manager --save --setopt=<repoid>.skip_if_unavailable=true

Cannot find a valid baseurl for repo: base/7/x86_64
```

An investigation and correction is required.

# Update Report

After testing the uploaded Vagrant box in Vagrant Cloud, the box was not working because it didn't have VirtualBox Guest Additions installed. This module is necessary for the allocator to copy the SSH keys into the machine (apart from this, this module allows shared folders and other functionalities):

```console
[2024-07-02 19:40:34] [INFO] ALLOCATOR: Creating instance at /tmp/wazuh-qa
[2024-07-02 19:40:35] [DEBUG] ALLOCATOR: No config provided. Generating from payload
[2024-07-02 19:40:35] [DEBUG] ALLOCATOR: Generating new key pair
[2024-07-02 19:40:38] [DEBUG] ALLOCATOR: Vagrantfile created. Creating instance.
[2024-07-02 19:40:38] [INFO] ALLOCATOR: Instance linux-centos-7-amd64-5028 created.
[2024-07-02 19:43:49] [ERROR] ALLOCATOR: Command failed: b'Vagrant was unable to mount VirtualBox shared folders. This is usually\nbecause the filesystem "vboxsf" is not available. This filesystem is\nmade available via the VirtualBox Guest Additions and kernel module.\nPlease verify that these guest additions are properly installed in the\nguest. This is not a bug in Vagrant and is usually caused by a faulty\nVagrant box. For context, the command attempted was:\n\nmount -t vboxsf -o uid=1000,gid=1000,_netdev vagrant /vagrant\n\nThe error output from the command was:\n\nmount: unknown filesystem type \'vboxsf\'\n'
[2024-07-02 19:43:49] [INFO] ALLOCATOR: Instance linux-centos-7-amd64-5028 started.
[2024-07-02 19:43:53] [INFO] ALLOCATOR: The inventory file generated at /tmp/dtt1-poc/linux-centos-7-vg/inventory.yaml
[2024-07-02 19:43:53] [INFO] ALLOCATOR: The track file generated at /tmp/dtt1-poc/linux-centos-7-vg/track.yaml
```


I rebuilt the machine again and I did the following steps:
- Download the `.iso` file from `https://download.virtualbox.org/virtualbox/${VIRTUALBOX_VERSION}/VBoxGuestAdditions_${VIRTUALBOX_VERSION}.iso`, where `VIRTUALBOX_VERSION` is the VirtualBox version of the host. 
- Mount the ISO file into a new folder.
- Make sure that the kernel and the kernel sources match. This is verified with the `uname -a` and `ls /usr/src/kernels`. If they do not match, it is necessary to update the system with `yum update` and install the kernel-devel and kernel-headers packages with the `yum install kernel-devel-<KERNEL_VERSION> kernel-headers-<KERNEL_VERSION>` command.
- Run the `ISO` file: `sh /mnt/VBoxLinuxAdditions.run`
- Verify that the Guest Additions is installed with the `lsmod | grep vbox` command, which must generate an output if it's installed correctly.

After this, it is necessary to replace the private Vagrant key (if was generated) and reboot the system.


The Vagrant box was built successfully and it is working correctly. Notice that the GuestAdditions warning is not shown in the vagrant output. Besides, the system was updated to CentOS 7.9.

```console
vagrant up                             
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'centos7-GA'...
==> default: Matching MAC address for NAT networking...
==> default: Setting the name of the VM: centos-test2_default_1720000645649_15989
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: 
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default: 
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Mounting shared folders...
    default: /vagrant => /home/davidcr01/centos-test2
 ~/centos-test2  vagrant ssh 
Last login: Wed Jul  3 18:50:27 2024 from 10.0.2.2
[vagrant@localhost ~]$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

[vagrant@localhost ~]$ cat /etc/centos-release
CentOS Linux release 7.9.2009 (Core)
[vagrant@localhost ~]$ lsmod | grep vbox
vboxsf                 81130  1 
vboxvideo              39963  1 
ttm                   100769  1 vboxvideo
drm_kms_helper        186531  1 vboxvideo
drm                   468454  4 ttm,drm_kms_helper,vboxvideo
vboxguest             357598  2 vboxsf
[vagrant@localhost ~]$ 
``

# Update Report

After testing the uploaded Vagrant box in Vagrant Cloud, the box was not working because it didn't have VirtualBox Guest Additions installed. This module is necessary for the allocator to copy the SSH keys into the machine (apart from this, this module allows shared folders and other functionalities):

```console
[2024-07-02 19:40:34] [INFO] ALLOCATOR: Creating instance at /tmp/wazuh-qa
[2024-07-02 19:40:35] [DEBUG] ALLOCATOR: No config provided. Generating from payload
[2024-07-02 19:40:35] [DEBUG] ALLOCATOR: Generating new key pair
[2024-07-02 19:40:38] [DEBUG] ALLOCATOR: Vagrantfile created. Creating instance.
[2024-07-02 19:40:38] [INFO] ALLOCATOR: Instance linux-centos-7-amd64-5028 created.
[2024-07-02 19:43:49] [ERROR] ALLOCATOR: Command failed: b'Vagrant was unable to mount VirtualBox shared folders. This is usually\nbecause the filesystem "vboxsf" is not available. This filesystem is\nmade available via the VirtualBox Guest Additions and kernel module.\nPlease verify that these guest additions are properly installed in the\nguest. This is not a bug in Vagrant and is usually caused by a faulty\nVagrant box. For context, the command attempted was:\n\nmount -t vboxsf -o uid=1000,gid=1000,_netdev vagrant /vagrant\n\nThe error output from the command was:\n\nmount: unknown filesystem type \'vboxsf\'\n'
[2024-07-02 19:43:49] [INFO] ALLOCATOR: Instance linux-centos-7-amd64-5028 started.
[2024-07-02 19:43:53] [INFO] ALLOCATOR: The inventory file generated at /tmp/dtt1-poc/linux-centos-7-vg/inventory.yaml
[2024-07-02 19:43:53] [INFO] ALLOCATOR: The track file generated at /tmp/dtt1-poc/linux-centos-7-vg/track.yaml
```


I rebuilt the machine again and I did the following steps:
- Download the `.iso` file from `https://download.virtualbox.org/virtualbox/${VIRTUALBOX_VERSION}/VBoxGuestAdditions_${VIRTUALBOX_VERSION}.iso`, where `VIRTUALBOX_VERSION` is the VirtualBox version of the host. 
- Mount the ISO file into a new folder.
- Make sure that the kernel and the kernel sources match. This is verified with the `uname -a` and `ls /usr/src/kernels`. If they do not match, it is necessary to update the system with `yum update` and install the kernel devel and kernel headers packages with the `yum install kernel-devel-<KERNEL_VERSION> kernel-headers-<KERNEL_VERSION>` command.
- Run the `ISO` file: `sh /mnt/VBoxLinuxAdditions.run`
- Verify that the Guest Additions is installed with the `lsmod | grep vbox` command, which must generate an output if it's installed correctly.

After this, it is necessary to replace the private Vagrant key (if was generated) and reboot the system.


The Vagrant box was built succesfully and it is working correctly. Notice that the GuestAdditions warning is not shown in the vagrant output. Besides, the system was updated to CentOS 7.9.

```console
> vagrant up                             
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'centos7-GA'...
==> default: Matching MAC address for NAT networking...
==> default: Setting the name of the VM: centos-test2_default_1720000645649_15989
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: 
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default: 
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Mounting shared folders...
    default: /vagrant => /home/davidcr01/centos-test2

> vagrant ssh 
Last login: Wed Jul  3 18:50:27 2024 from 10.0.2.2
[vagrant@localhost ~]$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

[vagrant@localhost ~]$ cat /etc/centos-release
CentOS Linux release 7.9.2009 (Core)
[vagrant@localhost ~]$ lsmod | grep vbox
vboxsf                 81130  1 
vboxvideo              39963  1 
ttm                   100769  1 vboxvideo
drm_kms_helper        186531  1 vboxvideo
drm                   468454  4 ttm,drm_kms_helper,vboxvideo
vboxguest             357598  2 vboxsf
[vagrant@localhost ~]$ 
```

After re-uploading the box to VagrantCloud, the Allocator is working correctly:
```console
> python3  modules/allocation/main.py --action create --provider vagrant --size micro --composite-name linux-centos-7-amd64 --inventory-output  /tmp/dtt1-poc/linux-centos-7-vg/inventory.yaml --track-output  /tmp/dtt1-poc/linux-centos-7-vg/track.yaml --instance-name linux-centos-7-amd64
[2024-07-03 12:51:04] [INFO] ALLOCATOR: Creating instance at /tmp/wazuh-qa
[2024-07-03 12:51:05] [DEBUG] ALLOCATOR: No config provided. Generating from payload
[2024-07-03 12:51:05] [DEBUG] ALLOCATOR: Generating new key pair
[2024-07-03 12:51:09] [DEBUG] ALLOCATOR: Vagrantfile created. Creating instance.
[2024-07-03 12:51:09] [INFO] ALLOCATOR: Instance linux-centos-7-amd64-7149 created.
[2024-07-03 12:52:32] [INFO] ALLOCATOR: Instance linux-centos-7-amd64-7149 started.
[2024-07-03 12:52:36] [INFO] ALLOCATOR: The inventory file generated at /tmp/dtt1-poc/linux-centos-7-vg/inventory.yaml
[2024-07-03 12:52:36] [INFO] ALLOCATOR: The track file generated at /tmp/dtt1-poc/linux-centos-7-vg/track.yaml
[2024-07-03 12:52:36] [INFO] ALLOCATOR: SSH connection successful.
[2024-07-03 12:52:36] [INFO] ALLOCATOR: Instance linux-centos-7-amd64-7149 created successfully.

> ssh -i /tmp/wazuh-qa/linux-centos-7-amd64-7149/instance_key 192.168.57.115
The authenticity of host '192.168.57.115 (192.168.57.115)' can't be established.
ED25519 key fingerprint is SHA256:lbiAUdup7OT5Ffrk0eXFccFpF03E59kDdrY+nZT2bYM.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? ^C
 ✘  ~/Wazuh/wazuh-qa/deployability  ↱ bug/5544-fix-centos7-allocator ±  ssh -i /tmp/wazuh-qa/linux-centos-7-amd64-7149/instance_key vagrant@192.168.57.115
The authenticity of host '192.168.57.115 (192.168.57.115)' can't be established.
ED25519 key fingerprint is SHA256:lbiAUdup7OT5Ffrk0eXFccFpF03E59kDdrY+nZT2bYM.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '192.168.57.115' (ED25519) to the list of known hosts.
Last login: Wed Jul  3 19:20:36 2024 from 10.0.2.2
[vagrant@localhost ~]$ sudo su
[root@localhost vagrant]# cat /etc/centos-release
CentOS Linux release 7.9.2009 (Core)
[root@localhost vagrant]# 
```

--- a/deployability/modules/allocation/static/specs/os.yml
+++ b/deployability/modules/allocation/static/specs/os.yml
@@ -61,8 +61,8 @@ vagrant:
     virtualizer: virtualbox
   # Centos
   linux-centos-7-amd64:
-    box: centos/7
-    box_version: 2004.01
+    box: vagrant_devel/centos-7
+    box_version: 1.1.0
     virtualizer: virtualbox
   linux-centos-7-ppc64:
     box: testing-centos-image

